tsl/src/remote/copy_fetcher.c

```diff
@@ -240,6 +241,66 @@ copy_data_check_header(StringInfo copy_data)
 	}
 }
 
+/*
+ * End the COPY after receiving EOF or canceling a query (e.g., due to a LIMIT
+ * being reached).
+ *
+ * This should be called after canceling a query, or, after reading all data,
+ * the file trailer, and getting an EOF return value.
+ */
+static void
+end_copy(CopyFetcher *fetcher)
+{
+	PGconn *conn = remote_connection_get_pg_conn(fetcher->state.conn);
+	PGresult *final_pgres = NULL;
+	PGresult *pgres = NULL;
+	/* Expect COMMAND OK if we received EOF. If the query was canceled (e.g.,
+	 * due to a LIMIT), expect FATAL_ERROR (query failed) */
+	ExecStatusType expected_status = fetcher->state.eof ? PGRES_COMMAND_OK : PGRES_FATAL_ERROR;
```

Comment:
In case of LIMIT, why are we setting FATAL_ERROR? We want to close early and with SUCCESS, no?

Reply:
This is not us. This is the error returned from the data node where the query is canceled. From `PQcancel()` docs:

" If the cancellation is effective, the current command will terminate early and return an error result."

Reply:
Actually, when we cancel, the docs says we can receive either COMMAND_OK or ERROR. For example, the query could have just completed when the canceled happened. 

I will tweak the code to account for this case and make things more clear.